### DISCUSSION TITLE
Std/Col request negotiation, request negotiations vs. respond to nego request.

### DISCUSSION THREAD

**Discussion by umutcakan** [yasserfarouk/scml#34] [Q&A] [answered]
Hey there,

I have several questions regarding the SCML21 std/col track.

1) Is there any 'call' order from World between request_negotiation or/and request_negotiations vs. respond_to_negotiation_request methods?

For example, if I call request_negotiation or negotiations in step() function, when do I receive the information regarding that agent/agents accepted my request, and when my respond_to_negotiation_request method is called? So that I will know we will negotiate the next step with which agents (if I'm not mistaken) and which agents I have not yet scheduled a negotiation with, so that I can accept their requests.

2) If they called simultaneously, is it possible to negotiate with 1 agent with multiple negotiators?

For example, if I send a negotiation request and the agent accepts it & sends one request back and I accept it (since I do not know if we already scheduled a negotiation or not).

3) Does request_for_negotiations return boolean depending on all agents' acceptance? 

For example, if I call request_negotiation in step() function, I know it returns true if the requested agent accepts this request since we are sending requests to 1 partner. But if we call request_negotiations, it also returns true, but with no / zero context (e.g. which agents accepted the request. I am assuming it returns true if all agents accepted my negotiation request. In this case, would not it be better to return with agent IDs and their acceptance status? For example, {'agent1': accept, 'agent2': reject} etc.

4) Is there any flow diagram for an agent ? 

I know there are diagrams for negotiation progress (negotiators - Mechanism etc). But I could not find one for an agent (e.g. at the start of a day (step), first, step function is called then respond_to_negotiation is called etc.)

5) Is it possible to have 2 negotiators in different steps in while using counterall method? (SyncController)

For example, is it possible to have 1 negotiator in step 3, while other one is in step 9? I know Mechanism breaks loop if there is a lock between agents that are using synccontroller, but other than that does SyncController really syncs the process?

Thanks!

**Comment by yasserfarouk** [comment, marked as answer]
Thank you for your interest in SCML and the excellent questions

> Is there any 'call' order from World between request_negotiation or/and request_negotiations vs. respond_to_negotiation_request methods?

- The call to `request_negotaition(s)` is synchronous meaning that it will return
  to you with the required information. Nevertheless, you can also just ignore this
	return value because you also receive a `on_negotiation_request_accepted()` and
	`on_negotiation_request_rejected()` callback for each negotiation you request.


> If they called simultaneously, is it possible to negotiate with 1 agent with
multiple negotiators?

- Yes you can but we have an upper limit on the number of negotiations you can
	engage in. There are three limits:
	```python
	negotiation_quota_per_step=float("inf"),
	negotiation_quota_per_simulation=float("inf"),
	n_concurrent_negs_between_partners=float("inf"),
	```
	For now all are infinite which means you can request any number of negotiations
	with any opponent. **Nevertheless, this has huge computational cost so we may
	restrict these numbers in the official competition.** When these restrictions
	are applied, you will simply receive a rejection for any extra request you make.
	Your agent can read these values in real-time from the settings in its awi.
	For example:
	```python
	self.awi.settings["n_concurrent_negs_between_partners"]
	```


> Does request_for_negotiations return boolean depending on all agents' acceptance?

- No. It returns true if **any** agent accepted to start a negotiation. It may
  be better to rely on `on_negotiation_request_accepted` and `on_negotiation_request_rejected`.


> In this case, would not it be better to return with agent IDs and their acceptance status? For example, {'agent1': accept, 'agent2': reject} etc.

- Actually, that makes sense. Nevertheless, we cannot modify it this year for
  backward compatibility as other agents already developed may depend on the fact
	that it returns a boolean giving true.
- The reason we opted for this design is that `request_negotaition` and `request_negotiations` can
  be converted into async calls easily if agents do not rely on their return value but
	on `on_negotiation_request_accepted` and `on_negotiation_request_rejected` calls. I guess
	that we should have made the return value `None` though.

> Is there any flow diagram for an agent ?

- The nearest thing to this is found in [ this video (starting 2:34) ](https://youtu.be/velp6T0dUno?t=154).

> Is it possible to have 2 negotiators in different steps in while using counterall method? (SyncController) For example, is it possible to have 1 negotiator in step 3, while other one is in step 9?

- I cannot be really sure 😅. I guess it may be possible in general but in SCML may be it does not happen.

> I know Mechanism breaks loop if there is a lock between agents that are using synccontroller, but other than that does SyncController really syncs the process?

- The meaning of `sync` here is just that the controller with **try** to collect all offers from all partners and present them all to your `counter_all` call. It may fail due to loops and in such 
cases it will present you with offers from a subset (which may have cardinality one) of your partners. It does not try to do any more syncing that that. Specifically, it does not try to sync the 
round number in all negotiations.

**Comment by umutcakan** [comment]
Thanks for your answers Yasser, it helped a lot! Sorry for the late response, I was working on several features :)

> * The call to `request_negotaition(s)` is synchronous meaning that it will return
>   to you with the required information. Nevertheless, you can also just ignore this
>   return value because you also receive a `on_negotiation_request_accepted()` and
>   `on_negotiation_request_rejected()` callback for each negotiation you request.

This is just great. In this case, I was wondering if World calls each agent's 'step' function and waits, or it just calls and moves to the next agent. Because if World calls and moves to the next agent, it is possible to simultaneously receive and send negotiation requests, if an agent designer sends their negotiation requests inside the step function (I am assuming this is the case most of the time).

And this situation can be problematic for this case you've mentioned. 

> * Yes you can but we have an upper limit on the number of negotiations you can
>   engage in. There are three limits:
>   ```python
>    negotiation_quota_per_step=float("inf"),
>    negotiation_quota_per_simulation=float("inf"),
>    n_concurrent_negs_between_partners=float("inf"),
>   ```

Because, if I receive and send requests at the same time, it is possible that I can have 2 negotiators for this agent at the next step unintentionally, and as an agent designer I do not want that all the time, specifically if I have a limited quota.


> * The nearest thing to this is found in [ this video (starting 2:34) ](https://youtu.be/velp6T0dUno?t=154).

This is just what I've asked for! I checked on your youtube channel, and the visualization part looks so promising! Can't wait to test it out.

I have several new questions 😅

1) Can I set the agent position before tournament starts? For example, I want it to be in the middle of the chain for a specific case so that I won't be getting exogenous contracts.

2) Is it possible to disable the default logger and use a new one? I am trying to implement a new logger (I am kinda new with this, might be a problem from my side), but World's logger throws an error like this:  

scml\lib\concurrent\futures\_base.py", line 384, in __get_result
ValueError: list.remove(x): x not in list

3) Is it possible to run a tournament with agents where each of them has 1 copy only? (I am not sure if this is the case for the real tournament) At the start, I was trying to print some statements on the console, but they are running parallel if I am not mistaken. Is it possible to run only 1 tournament session? I have set n_configs = 1, max_n_world_per_config=1, and n_runs_per_world=1, but system still creates multiple agents per session.

**Comment by yasserfarouk** [comment]
> I was wondering if World calls each agent's 'step' function and waits, or it just calls and moves to the next agent. Because if World calls and moves to the next agent, it is possible to simultaneously receive and send negotiation requests, if an agent designer sends their negotiation requests inside the step function (I am assuming this is the case most of the time).

All callbacks in SCML (and NegMAS) are atomic. You will never receive one of your callbacks while another is in progress. Moreover, for the version we are using in the tournaments (the one on github), each world simulation is serialized in the sense that no callbacks even from different agents can run simultaneously.

> Can I set the agent position before tournament starts? For example, I want it to be in the middle of the chain for a specific case so that I won't be getting exogenous contracts.

For a tournament no. Nevertheless, you can easily arrange that by creating worlds directly instead of tournaments. See the method `generate()` in each world (See [this](http://www.yasserm.com/scml/scml2020docs/api/scml.oneshot.SCML2020OneShotWorld.html#scml.oneshot.SCML2020OneShotWorld.generate) )

> Is it possible to disable the default logger and use a new one? I am trying to implement a new logger (I am kinda new with this, might be a problem from my side)

- Do you mean the world logger? If so, you can disable it by passing `log_to_file=False` (and doing the same for all `log_* variables` in the world constructor (See [the world constructor](http://www.yasserm.com/negmas/api/negmas.situated.World.html) ). To implement your own logger, you can just hook into the event system in NegMAS using an [event logger](http://www.yasserm.com/negmas/api/negmas.events.EventLogger.html?highlight=logger#negmas.events.EventLogger) which will receive all the events that are raised during the simulation. You can do whatever you want with them. Check [this](http://www.yasserm.com/negmas/modules/situated.html#monitoring-a-simulation) for a *partial* list of the events you can monitor and their meaning.

 - If you want a logger that logs information available to your agent, you can use the `log*_agent` methods (check [this](http://www.yasserm.com/negmas/api/negmas.situated.World.html#negmas.situated.World.logdebug_agent) ) which allows your agent to log into its own file which will be stored under `_agent_logs` in the world logs. Check [this tutorial](http://www.yasserm.com/scml/scml2020docs/tutorials/04.logs_and_stats.html) about logging in general.


> World's logger throws an error like this:
```
scml\lib\concurrent\futures_base.py", line 384, in __get_result
ValueError: list.remove(x): x not in list
```

It is hard to understand what is happening by looking at this. If you can create a small code fragment that creates it, I can be of help

> Is it possible to run a tournament with agents where each of them has 1 copy only?

You can control the number of agent per competitor but not for non-competitors (because the system needs to fill all other factories not taken by competitors). One possibility is to have a number of competitors that is enough to cover all factories and using `anac2021_std` or `anac2021_oneshot` from `scml.utils`. These guarantee that there is one agent per competitor. Another possibility is to directly call `SCML2020World.generate()`/`SCML2020OneShotWorld.generate()` and you can control the placement of agents as well as their numbers completely.

> Is it possible to run only 1 tournament session? 

I think you mean running one world simulation. This is what `generate()` does. If you mean running the tournament serially, you can do that by passing --serial to the CLI
